examples/ticks_and_spines/date_precision_and_epochs.py

```diff
@@ -0,0 +1,154 @@
+"""
+=========================
+Date Precision and Epochs
+=========================
+
+Matplotlib can handle `.datetime` objects and `numpy.datetime64` objects using
+a unit converter that recognizes these dates and converts them to floating
+point numbers.
+
+Before Matplotlib 3.3, the default for this conversion returns a float that was
+days since "0000-12-31T00:00:00".  As of Matplotlib 3.3, the default is
+days from "1970-01-01T00:00:00".  This allows more resolution for modern
+dates.  "2020-01-01" with the old epoch converted to 730120, and a 64-bit
+floating point number has a resolution of 2^{-52}, or approximately
+14 microseconds, so microsecond precision was lost.  With the new default
+epoch "2020-01-01" is 10957.0, so the achievable resolution is 0.21
+microseconds.
+
+"""
+import datetime
+import dateutil
+import numpy as np
+
+import matplotlib
+import matplotlib.pyplot as plt
+import matplotlib.dates as mdates
+
+#############################################################################
+# Datetime
+# --------
+#
+# Python `.datetime` objects have microsecond resolution, so with the
+# old default matplotlib dates could not round-trip full-resolution datetime
+# objects.
+
+old_epoch = '0000-12-31T00:00:00'
+new_epoch = '1970-01-01T00:00:00'
+
+mdates._reset_epoch()  # Don't do this.  Just being done for this tutorial.
+mdates.set_epoch(old_epoch)  # old epoch (pre MPL 3.3)
+
+date1 = datetime.datetime(2000, 1, 1, 0, 10, 0, 12,
+                          tzinfo=dateutil.tz.gettz('UTC'))
+mdate1 = mdates.date2num(date1)
+print('Before Roundtrip: ', date1, 'Matplotlib date:', mdate1)
+date2 = mdates.num2date(mdate1)
+print('After Roundtrip:  ', date2)
+
+#############################################################################
+# Note this is only a round-off error, and there is no problem for
+# dates closer to the old epoch:
+
+date1 = datetime.datetime(10, 1, 1, 0, 10, 0, 12,
+                          tzinfo=dateutil.tz.gettz('UTC'))
+mdate1 = mdates.date2num(date1)
+print('Before Roundtrip: ', date1, 'Matplotlib date:', mdate1)
+date2 = mdates.num2date(mdate1)
+print('After Roundtrip:  ', date2)
+
+#############################################################################
+# If a user wants to use modern dates at microsecond precision, they
+# can change the epoch.  However, note that calling `.set_epoch` yields an
+# error here.  That is because mixing epochs in a script is confusing, so
+# we attach a sentinel that doesn't let it be changed after any dates
+# code has been run.  In order to run `~.set_epoch` it must be called before
+# any date code has been executed (i.e. near the top of a script).
```

Comment:
```suggestion
# can change the epoch using `~.set_epoch`.  However, the epoch has to be
# set before any date operations to prevent confusion between different
#  epochs. Trying to change the epoch later will raise a `RuntimeError`.
```